[Bugfix] Mistral tool parser streaming update

avigny opened the pull request:
## Purpose

Fixes https://github.com/vllm-project/vllm/issues/13622
Fixes https://github.com/vllm-project/vllm/issues/17585
Fixes https://github.com/vllm-project/vllm/issues/20028

This PR is similar to #16096 (hermes tool parser)

## In summary

### Repairs tool call in streaming mode for (older) models with tokenizer version <v11
The model output is incrementaly parsed with `ijson` emitting events used to know what is being streamed (what part of the tool call). for more details see [`_extract_tool_calls_streaming_pre_v11_tokenizer`](https://github.com/avigny/vllm/blob/e63ee21bde43e3e40c294e2aac9c7dd1957be83d/vllm/entrypoints/openai/tool_parsers/mistral_tool_parser.py#L360)
Quick unit tests added in `tests/tool_use/test_mistral_tool_parser.py` see [`test_extract_tool_calls_streaming_pre_v11_tokenizer`](https://github.com/avigny/vllm/blob/e63ee21bde43e3e40c294e2aac9c7dd1957be83d/tests/tool_use/test_mistral_tool_parser.py#L444)

### Adds support for tool calls in streaming mode for recent models (tokenizer version >=v11)
See [`_extract_tool_calls_streaming`](https://github.com/avigny/vllm/blob/e63ee21bde43e3e40c294e2aac9c7dd1957be83d/vllm/entrypoints/openai/tool_parsers/mistral_tool_parser.py#L226) for implementation details
Test added for `mistralai/Mistral-Small-3.2-24B-Instruct-2506` in `tests/tool_use/test_mistral_tool_parser.py`
Quick unit tests added in `tests/tool_use/test_mistral_tool_parser.py` see [`test_extract_tool_calls_streaming`
](https://github.com/avigny/vllm/blob/e63ee21bde43e3e40c294e2aac9c7dd1957be83d/tests/tool_use/test_mistral_tool_parser.py#L523)

## Test Plan

I've added a test file `tests/tool_use/test_mistral_tool_parser.py` for easy and fast testing. This file works similarly as the existing `tests/tool_use/test_jamba_tool_parser.py`.

This tests the parsing functions with a mocked model output. It allows toeasily test edge cases.

Use `pytest tests/tool_use/test_mistral_tool_parser.py` to run this test file.

Test added for `mistralai/Mistral-Small-3.2-24B-Instruct-2506` in `tests/tool_use/test_mistral_tool_parser.py`

## (Optional) Documentation Update
I believe no documentation update is needed
<!--- pyml disable-next-line no-emphasis-as-heading -->

avigny committed:
Testing mistral tool parser
Tests are similar as the ones added for Jamba models in https://github.com/vllm-project/vllm/pull/9154

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
Update streaming tool parser for mistral

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
Removing unneeded check

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
repair ruff pre-commit

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny commented:
@hibukipanim I did run the test you provided in your issue description https://github.com/vllm-project/vllm/issues/17585#issue-3035811793 and got the following output:

```python
ChoiceDeltaToolCall(index=0, id='j6OY9szTS', function=ChoiceDeltaToolCallFunction(arguments=None, name='mcp_confluence'), type='function')
ChoiceDeltaToolCall(index=0, id=None, function=ChoiceDeltaToolCallFunction(arguments='{"', name=None), type=None)
ChoiceDeltaToolCall(index=0, id=None, function=ChoiceDeltaToolCallFunction(arguments='query', name=None), type=None)
ChoiceDeltaToolCall(index=0, id=None, function=ChoiceDeltaToolCallFunction(arguments='":', name=None), type=None)
ChoiceDeltaToolCall(index=0, id=None, function=ChoiceDeltaToolCallFunction(arguments=' "', name=None), type=None)
ChoiceDeltaToolCall(index=0, id=None, function=ChoiceDeltaToolCallFunction(arguments='co', name=None), type=None)
ChoiceDeltaToolCall(index=0, id=None, function=ChoiceDeltaToolCallFunction(arguments='ffee', name=None), type=None)
ChoiceDeltaToolCall(index=0, id=None, function=ChoiceDeltaToolCallFunction(arguments='",', name=None), type=None)
ChoiceDeltaToolCall(index=0, id=None, function=ChoiceDeltaToolCallFunction(arguments=' "', name=None), type=None)
ChoiceDeltaToolCall(index=0, id=None, function=ChoiceDeltaToolCallFunction(arguments='limit', name=None), type=None)
ChoiceDeltaToolCall(index=0, id=None, function=ChoiceDeltaToolCallFunction(arguments='":', name=None), type=None)
ChoiceDeltaToolCall(index=0, id=None, function=ChoiceDeltaToolCallFunction(arguments=' ', name=None), type=None)
ChoiceDeltaToolCall(index=0, id=None, function=ChoiceDeltaToolCallFunction(arguments='1', name=None), type=None)
ChoiceDeltaToolCall(index=0, id=None, function=ChoiceDeltaToolCallFunction(arguments='}', name=None), type=None)
```
It seems to fix your issue.
Please let me know If I missed something.

PedroMiolaSilva commented:
@avigny hey!

I've being trying to test your solution but with no success. This is what I'm doing:

```
source ../.env
export MODEL_ID=unsloth/Mistral-Small-3.2-24B-Instruct-2506-FP8
export MODEL_ID_PORT=8000
export MODEL_ID_GPU=0

docker run \
--runtime nvidia \
-e VLLM_USE_V1=1 \
--gpus all \
--ipc=host \
-p "${MODEL_ID_PORT}:8000" \
--env "HUGGING_FACE_HUB_TOKEN=${HUGGING_FACE_HUB_TOKEN}" \
--env "HF_HUB_OFFLINE=0" \
-v "${HF_HOME}:/root/.cache/huggingface" \
-v "./mistral_tool_parser.py:/usr/local/lib/python3.12/dist-packages/vllm/entrypoints/openai/tool_parsers/mistral_tool_parser.py" \
vllm/vllm-openai:latest \
-v "$(pwd):/app" \
--model ${MODEL_ID} \
--tool-call-parser mistral \
--chat-template /app/template.jinja
--enable-auto-tool-choice \
--limit-mm-per-prompt 'image=1' \
--tokenizer_mode mistral \
--config_format mistral \
--load_format mistral \
--max-model-len 64000 \
--gpu-memory-utilization 0.8
```
Where template.jinja is [this one](https://github.com/vllm-project/vllm/blob/main/examples/tool_chat_template_mistral3.jinja) and mistral_tool_parser.py is the one that you've created.

I'm using this test request:

```
curl -X POST \
   http://localhost:8000/v1/chat/completions \
   -H "Content-Type: application/json" \
   -d '{
   "model": "unsloth/Mistral-Small-3.2-24B-Instruct-2506-FP8",
   "messages": [
    {"role":"system","content":"You have access to the weather tool. You should call this tool when you think it makes sense"},
     {"role": "user", "content": "What'\''s the weather in New York?"}
   ],
   "tools": [
     {
       "type": "function",
       "function": {
         "name": "get_weather",
         "description": "Get the current weather in a given location",
         "parameters": {
           "type": "object",
           "properties": {
             "location": {
               "type": "string",
               "description": "The city and state, e.g. San Francisco, CA"
             }
           },
           "required": ["location"]
         }
       }
     }
   ]
 }'
```

When I set stream to false, I'm getting this response:
```
{"id":"chatcmpl-0dc2b75406114cbcb4f95735ccfdb094","object":"chat.completion","created":1751490167,"model":"unsloth/Mistral-Small-3.2-24B-Instruct-2506-FP8","choices":[{"index":0,"message":{"role":"assistant","reasoning_content":null,"content":"[TOOL_CALLS]get_weather{\"location\": \"New York, NY\"}","tool_calls":[]},"logprobs":null,"finish_reason":"stop","stop_reason":null}],"usage":{"prompt_tokens":112,"total_tokens":127,"completion_tokens":15,"prompt_tokens_details":null},"prompt_logprobs":null,"kv_transfer_params":null}
```

And this error:
```
ERROR 07-02 14:00:22 [mistral_tool_parser.py:160] Error in extracting tool call from response.
ERROR 07-02 14:00:22 [mistral_tool_parser.py:160] Traceback (most recent call last):
ERROR 07-02 14:00:22 [mistral_tool_parser.py:160]   File "/usr/local/lib/python3.12/dist-packages/vllm/entrypoints/openai/tool_parsers/mistral_tool_parser.py", line 131, in extract_tool_calls
ERROR 07-02 14:00:22 [mistral_tool_parser.py:160]     function_call_arr = json.loads(tool_content)
ERROR 07-02 14:00:22 [mistral_tool_parser.py:160]                         ^^^^^^^^^^^^^^^^^^^^^^^^
ERROR 07-02 14:00:22 [mistral_tool_parser.py:160]   File "/usr/lib/python3.12/json/__init__.py", line 346, in loads
ERROR 07-02 14:00:22 [mistral_tool_parser.py:160]     return _default_decoder.decode(s)
ERROR 07-02 14:00:22 [mistral_tool_parser.py:160]            ^^^^^^^^^^^^^^^^^^^^^^^^^^
ERROR 07-02 14:00:22 [mistral_tool_parser.py:160]   File "/usr/lib/python3.12/json/decoder.py", line 338, in decode
ERROR 07-02 14:00:22 [mistral_tool_parser.py:160]     obj, end = self.raw_decode(s, idx=_w(s, 0).end())
ERROR 07-02 14:00:22 [mistral_tool_parser.py:160]                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ERROR 07-02 14:00:22 [mistral_tool_parser.py:160]   File "/usr/lib/python3.12/json/decoder.py", line 356, in raw_decode
ERROR 07-02 14:00:22 [mistral_tool_parser.py:160]     raise JSONDecodeError("Expecting value", s, err.value) from None
ERROR 07-02 14:00:22 [mistral_tool_parser.py:160] json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
ERROR 07-02 14:00:22 [mistral_tool_parser.py:160] 
ERROR 07-02 14:00:22 [mistral_tool_parser.py:160] During handling of the above exception, another exception occurred:
ERROR 07-02 14:00:22 [mistral_tool_parser.py:160] 
ERROR 07-02 14:00:22 [mistral_tool_parser.py:160] Traceback (most recent call last):
ERROR 07-02 14:00:22 [mistral_tool_parser.py:160]   File "/usr/local/lib/python3.12/dist-packages/vllm/entrypoints/openai/tool_parsers/mistral_tool_parser.py", line 137, in extract_tool_calls
ERROR 07-02 14:00:22 [mistral_tool_parser.py:160]     raw_tool_call = self.tool_call_regex.findall(tool_content)[0]
ERROR 07-02 14:00:22 [mistral_tool_parser.py:160]                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^
ERROR 07-02 14:00:22 [mistral_tool_parser.py:160] IndexError: list index out of range
```

When I set stream=true, I dont receive any errors, but the response does not have tool calls:
```
data: {"id":"chatcmpl-028934e8ee754938943457f631313546","object":"chat.completion.chunk","created":1751490269,"model":"unsloth/Mistral-Small-3.2-24B-Instruct-2506-FP8","choices":[{"index":0,"delta":{"role":"assistant","content":""},"logprobs":null,"finish_reason":null}]}
```

Am I doing something wrong here?

avigny commented:
Yes you're both right!
I believe I did branch out and started working on my fix before the changes introduced by #19193 which introduced the use of `fn_name_regex` from the model tokenizer.
I'll try to port this to the `extract_tool_calls_streaming` method.

Thanks for finding this!

gaby commented:
Any update on getting this merge?

DarkLight1337 commented:
cc @aarnphm

PedroMiolaSilva reviewed (commented):
I think replacing lines 127:139 with this below will fix it for non-streaming:
```try:
            #First, use the tool call token to split, and we discard the first item, because it is empty
            raw_tool_calls = model_output.split(self.bot_token)[1:] 
            function_call_arr = []
            for raw_tool_call in raw_tool_calls:
                tool_name = raw_tool_call.split("{")[0]
                tool_arguments_begin = raw_tool_call.find("{")
                tool_arguments = raw_tool_call[tool_arguments_begin:]
                function_call_arr.append({
                                        "name": tool_name,
                                        "arguments": json.loads(tool_arguments)
                })
```

avigny committed:
Merge branch 'vllm-project:main' into mistral-tool-parser-streaming-update

avigny committed:
Adding SPDX header

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
Update non streaming tests with v11 tokenizer and tool call format

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
Updating tests for new v11 mistral tokenizer and new tool call format in the model output

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
Adding support for the new tool call format in mistral models

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
Merge branch 'vllm-project:main' into mistral-tool-parser-streaming-update

avigny committed:
Comment update

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny commented on vllm/entrypoints/openai/tool_parsers/mistral_tool_parser.py:
```diff
@@ -185,185 +203,357 @@ def extract_tool_calls_streaming(
 
         # if the tool call token ID IS in the tokens generated so far, that
         # means we're parsing as tool calls now
+        if _is_fn_name_regex_support(self.model_tokenizer):
```
Naming of this function could be modified to be more readable as the streaming function does not use this .

Basically this is used here to know if the tool calls should be parsed with :
- the old parser (tool call like `[TOOL_CALLS][{"name":"add" , "arguments":{"a": 3, "b": 4} } ]`)
- the new parser (tool call like `[TOOL_CALLS]add{"a": "3", "b": "4"}`)

avigny committed:
Repairing multi tool calls in non streaming mode

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny commented on tests/tool_use/test_mistral_tool_parser.py:
```diff
@@ -0,0 +1,528 @@
+        #                                            "b": 6
+        #                                        })))
+        #     ],
+        #     None) # Was already broken
```
should be repaired by [138cef3](https://github.com/vllm-project/vllm/pull/19425/commits/138cef34528c9f67c0a06724673b6c1a358d260c)

avigny committed:
CI repair

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
Adding comment

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
Simplify arguments of `_extract_tool_calls_streaming`

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
Repair Test

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
Streaming tool parsing improved for old pre v11 tokenizer
re-using @sjuxax `StreamingState` from https://github.com/vllm-project/vllm/pull/20503

Co-authored-by: Jeff Cook <jeff@jeffcook.io>
Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny commented:
/gemini review

avigny committed:
Merge remote-tracking branch 'upstream/main' into mistral-tool-parser-streaming-update

# Conflicts:
#	requirements/common.txt

avigny committed:
repair doc

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
Ensuring function name is sent in a single chunk

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny commented:
@sfbemerk 
This should be repaired in a338dc1
Function names should be sent once the name has completely been generated.

avigny committed:
Set `self.prev_tool_call_arr` after generating the first DeltaToolCall

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
Pre-commit formatting

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
Merge remote-tracking branch 'upstream/main' into mistral-tool-parser-streaming-update

# Conflicts:
#	requirements/common.txt

sfbemerk commented:
Thank a lot for the quick update! I can confirm that the function name is now streamed in only one chunk. 
There is still a difference to the other tool parsers, and also to the previous behavior for Mistral-Small-3.1:
What I see now is
```
"delta": {"role": "assistant", "content": "", "tool_calls": [{"id": "4189fdfaG", "index": 0, "type": "function", "function": null}]}
"delta": {"role": "assistant", "content": "", "tool_calls": []}
"delta": {"role": "assistant", "content": "", "tool_calls": []}
"delta": {"role": "assistant", "content": "", "tool_calls": []}
"delta": {"role": "assistant", "content": "", "tool_calls": [{"id": null, "index": 0, "type": "function", "function": {"name": "Wikipedia", "arguments": "{\""}}]}
```
where the first chunk with `"function": null` is different from other parsers. 
Compare with the output for Mistral-Small-3.1 and the old mistral tool parser:
```
"delta":{"tool_calls":[{"id":"Un9uQOzXu","type":"function","index":0,"function":{"name":"Wikipedia"}}]}
```
Again, I don't know which client library deals with "null" functions in which way. If you could defer streaming the entire first tool call chunk until the function name is known, this would at least make it consistent with the other parsers.

avigny committed:
Avoid returning empty DeltaMessages

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
Renaming `self.current_tool_id` to `self.current_tool_index`

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
Sending the tool id only in the first non empty chunk

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny commented:
The tool parser for old (pre v11 tokenizer) models should now
- not return empty deltas like `"delta": {"role": "assistant", "content": "", "tool_calls": []}`
- return the tool id with the first non empty chunk

I'm looking into doing this also for newer models' tool parser

avigny committed:
Avoiding returning empty delta tools and sending tool id with first non empty chunk

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
Pre commit reformatting

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

sfbemerk commented:
thank you, now it works fine for me!

avigny committed:
Variable renaming to repair pre-commit

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

DarkLight1337 commented on requirements/common.txt:
```diff
@@ -48,4 +48,5 @@ scipy # Required for phi-4-multimodal-instruct
 ninja # Required for xgrammar, rocm, tpu, xpu
 pybase64 # fast base64 implementation
 cbor2 # Required for cross-language serialization of hashable objects
+ijson # Required for mistral streaming tool parser
```
I don't see this being used in vLLM codebase, could you show the code that requires it?

avigny replied:
Yes this is not currently being used in vLLM code. I found it being used in #16096 and It seemed to be a good way to parse incrementally the json returned by the model in streaming mode.

I used it in `vllm/entrypoints/openai/tool_parsers/mistral_tool_parser.py`

### Sending parts of the json to be parsed
https://github.com/avigny/vllm/blob/59bf8fa8f0e0cea508837b55371a4ec86093cd4a/vllm/entrypoints/openai/tool_parsers/mistral_tool_parser.py#L431

### Catching ijson's events to update the streaming state
https://github.com/avigny/vllm/blob/59bf8fa8f0e0cea508837b55371a4ec86093cd4a/vllm/entrypoints/openai/tool_parsers/mistral_tool_parser.py#L320-L344

DarkLight1337 reviewed (approved):
LGTM

DarkLight1337 commented:
Tool use tests failing, PTAL

avigny committed:
Return an empty DeltaMessage so that finish_reason gets set

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny commented:
The tests
- `tool_use/test_parallel_tool_calls.py::test_parallel_tool_calls[mistral]`
- `tool_use/test_tool_calls.py::test_tool_call_and_choice[mistral]`

were failing because the finish reason was not sent.
I did the same trick as done here
https://github.com/vllm-project/vllm/blob/2836dd73f13015ee386c544760ca0d16888203f3/vllm/entrypoints/openai/tool_parsers/llama4_pythonic_tool_parser.py#L189-L191
to repair this.

Also Do you think it would be worth adding a new model to test the tool_use against?
What about a newer model like `mistralai/Mistral-Small-3.2-24B-Instruct-2506` that would use the new tool call style.
It would need to have a model tokenizer version >= 11
cf https://github.com/avigny/vllm/blob/037653725ee15e9ab047918b86c773a79cabddb1/vllm/entrypoints/openai/tool_parsers/mistral_tool_parser.py#L59-L61

DarkLight1337 commented:
Yes, let's add it

avigny committed:
Adding tools tests for `mistralai/Mistral-Small-3.2-24B-Instruct-2506` which uses the more recent mistral tool call format

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
Repair finish reason for new mistral models

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
Prevent missing `finish_reason` if the max_tokens limits cuts you off in the middle of the buffering.

change copied from 5fd8448e02c472586afe3be28bfb4e79415e0cad

Co-authored-by: sfbemerk <benjaminmerkel@mail.de>
Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
Only returning an empty `DeltaMessage` when tool parsing is finished

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny commented:
@DarkLight1337 I've added in f51deb1e53a2ab7690094c6f15dd5ffb32897f0f a config for testing tool calls on `mistralai/Mistral-Small-3.2-24B-Instruct-2506`

Also I've discovered that the `"--ignore-patterns=\"consolidated.safetensors\""` in
https://github.com/vllm-project/vllm/blob/9e0726e5bfd201fa2c9209e3997d24c72ecc3b13/tests/tool_use/utils.py#L123
is not taken into account when the models are downloaded in 
https://github.com/vllm-project/vllm/blob/9e0726e5bfd201fa2c9209e3997d24c72ecc3b13/tests/tool_use/conftest.py#L47
(not a big deal but worth mentioning)

avigny committed:
Merge branch 'vllm-project:main' into mistral-tool-parser-streaming-update

DarkLight1337 commented:
The tool use test is failing, PTAL again

avigny committed:
Repair `mistral-small-3.2` config for tests

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
Merge remote-tracking branch 'origin/mistral-tool-parser-streaming-update' into mistral-tool-parser-streaming-update

avigny committed:
Merge remote-tracking branch 'upstream/main' into mistral-tool-parser-streaming-update

avigny committed:
Running tool_use tests on 4 GPUs (L4) to be able to run the `mistralai/Mistral-Small-3.2-24B-Instruct-2506` model

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
fix pre commit

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
Merge branch 'vllm-project:main' into mistral-tool-parser-streaming-update

avigny committed:
Merge branch 'vllm-project:main' into mistral-tool-parser-streaming-update

avigny committed:
Merge branch 'vllm-project:main' into mistral-tool-parser-streaming-update

avigny committed:
Merge branch 'vllm-project:main' into mistral-tool-parser-streaming-update

avigny committed:
Merge remote-tracking branch 'upstream/main' into mistral-tool-parser-streaming-update

# Conflicts:
#	requirements/docs.txt

avigny commented:
hi @PedroMiolaSilva (sorry for the response delay)
I believe this is more of an issue from `mistral-common` or your client:

mistral-common (falsely) treats you message as a normal message with content because `message.content` is a string (and of course expects a non empty string)
https://github.com/mistralai/mistral-common/blob/38ab6d4b15e67a8d284c1d98253f989e52819320/src/mistral_common/tokens/tokenizers/instruct.py#L1029-L1030

The safe and easiest fix IMO is to discard your `message.content` if it's empty... (it works with a non empty content from my experience)

avigny committed:
Merge branch 'vllm-project:main' into mistral-tool-parser-streaming-update

avigny committed:
Merge branch 'vllm-project:main' into mistral-tool-parser-streaming-update

avigny committed:
Merge branch 'vllm-project:main' into mistral-tool-parser-streaming-update

ndebeiss commented:
Hello
We are using tool calls streaming with Mistral, and we have some issues.
I created a Docker with the code of 0.10.1 rebased with the code of that branch.
When debugging the method chat_completion_stream_generator, we observe that characters or pieces of text are doubled.
I think that comes from that code added recently :

```python
                    else:
                        # Update for comprehensive logging even in simple case
                        assert previous_texts is not None
                        previous_texts[i] += delta_text
```

It looks like it has been added for logging reasons : feat: Add --enable-log-outputs flag for logging model generations (#20707)

When removing that code, it is better but we have partial reasons coming : 

```
data: {"id":"chatcmpl-ebdadfec27834e0bb154b5e33a060038","object":"chat.completion.chunk","created":1755619510,"model":"mistralai/Mistral-Nemo-Instruct-2407","choices":[{"index":0,"delta":{"tool_calls":[{"id":"chatcmpl-tool-b5736bd715e84ef6b424965d15d09dad","type":"function","index":0,"function":{"name":"substractor","arguments":"{\""}}]},"logprobs":null,"finish_reason":null}]}

data: {"id":"chatcmpl-ebdadfec27834e0bb154b5e33a060038","object":"chat.completion.chunk","created":1755619510,"model":"mistralai/Mistral-Nemo-Instruct-2407","choices":[{"index":0,"delta":{"tool_calls":[{"index":0,"function":{"name":null,"arguments":"a"}}]},"logprobs":null,"finish_reason":null}]}

data: {"id":"chatcmpl-ebdadfec27834e0bb154b5e33a060038","object":"chat.completion.chunk","created":1755619510,"model":"mistralai/Mistral-Nemo-Instruct-2407","choices":[{"index":0,"delta":{"tool_calls":[{"index":0,"function":{"name":null,"arguments":"\":"}}]},"logprobs":null,"finish_reason":null}]}
```

maybe it is better not to partial load json until we find a ] or a ] ? with that kind of code :

```python
        if current_text[-1] not in [']', '}']:
            return None, function_name_returned
```

thanks for all that work !!

PedroMiolaSilva commented:
Hey guys, any news here?

hmellor commented on requirements/docs.txt:
Please merge from main, most requirements for the docs build have been removed now. This additional requirement could probably also be removed.

avigny committed:
Merge remote-tracking branch 'upstream/main' into mistral-tool-parser-streaming-update

# Conflicts:
#	requirements/docs.txt

avigny committed:
Merge branch 'vllm-project:main' into mistral-tool-parser-streaming-update

avigny committed:
Merge branch 'main' into mistral-tool-parser-streaming-update

avigny committed:
Merge remote-tracking branch 'upstream/main' into mistral-tool-parser-streaming-update

# Conflicts:
#	vllm/entrypoints/openai/serving_chat.py

avigny committed:
Merge remote-tracking branch 'origin/mistral-tool-parser-streaming-update' into mistral-tool-parser-streaming-update

avigny committed:
Merge branch 'main' into mistral-tool-parser-streaming-update

avigny commented:
[Magistral-Small](https://huggingface.co/mistralai/Magistral-Small-2509) and [Devstrall-small](https://huggingface.co/mistralai/Devstral-Small-2507) both use the new tekken v13 with a new tool call format...
For these models the tool calls look like :
```
[TOOL_CALLS]{"function": {"name": "function1", "arguments": {"arg1": "foo", "arg2": "bar"}}, {"function": {"name": "function2", "arguments": {"foobar": 1}}}
```

This was reported in https://github.com/vllm-project/vllm/issues/23180, and at the current state, this PR does not solve it.

I'm wondering if I should try and add a fix for these new models in the current PR and potentially make it even bigger or if I should open a separate PR only for the new models support

@DarkLight1337 or @aarnphm what do you suggest?

bbrowning commented on vllm/entrypoints/openai/tool_parsers/mistral_tool_parser.py:
```diff
@@ -185,185 +209,362 @@ def extract_tool_calls_streaming(
+        content = None
+        delta_tool_calls: list[DeltaToolCall] = []
+        current_tool_call: DeltaToolCall = DeltaToolCall(
+            index=self.current_tool_id, )
```
Right here we're missing a `type="function"` to make this a valid `DeltaToolCall` when returned back to the clients.

avigny replied:
Done in [6266dcf](https://github.com/vllm-project/vllm/pull/19425/commits/6266dcf060c98cebb927b71420a071f304d5fba0)

avigny committed:
Set type="function" to DeltaToolCall

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
Merge remote-tracking branch 'upstream/main' into mistral-tool-parser-streaming-update

# Conflicts:
#	tests/tool_use/utils.py
#	vllm/entrypoints/openai/tool_parsers/mistral_tool_parser.py

avigny committed:
Running ruff on `test_mistral_tool_parser.py`

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
Merge branch 'main' into mistral-tool-parser-streaming-update

avigny committed:
Repairing imports in `test_mistral_tool_parser.py`

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
Added check that `[TOOL_CALLS]` is a single bot token and not textual tokens

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
Merge remote-tracking branch 'upstream/main' into mistral-tool-parser-streaming-update

# Conflicts:
#	vllm/entrypoints/openai/tool_parsers/mistral_tool_parser.py

avigny committed:
Removing Optional and Union typing
same as https://github.com/vllm-project/vllm/pull/26633

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
Typos

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
Try-Catch the streaming parsing

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

bbrowning commented:
@avigny I didn't review all of the commits from scratch again, but the latest commits checking just for the bot_token_id (vs text) and error handling look good to me. Thanks for getting this in good shape!!

hmellor commented on .buildkite/test-pipeline.yaml:
```diff
@@ -553,6 +553,7 @@ steps:
 - label: OpenAI-Compatible Tool Use # 23 min
   timeout_in_minutes: 35
   mirror_hardwares: [amdexperimental]
+  num_gpus: 4
```
This will make this job significantly more expensive

avigny committed:
Removing tool test on `mistralai/Mistral-Small-3.2-24B-Instruct-2506`

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
Merge branch 'main' into mistral-tool-parser-streaming-update

avigny committed:
Revert "Removing tool test on `mistralai/Mistral-Small-3.2-24B-Instruct-2506`"

This reverts commit 58512a6e75e7684b0ef52ec334806ae76c21a12d.

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
Making `mistral-small-3.2` test not run by CI

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
Repair `mistral-small-3.2` test as mistral tokenizer cannot take custom chat template

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
Set `supports_parallel` to `True ` for both mistral models

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

alew3 commented:
@avigny I can see you are putting in a lot of work getting mistral tool calling to work! Thanks for all your effort! Can you update what the current status is?

patrickvonplaten reviewed (approved):
Very cool!

alew3 commented:
> This has taken so long that mistral/small is barely relevant nowadays.

I think it's still relevant! On the vllm world, most <32B models have some kind of issue with tool calling. Currently,  gpt-oss-20B (on the nightly docker)  is the only small model that works well for me with tool calling (stream/non stream).

alew3 commented:
Hey folks. A bunch of us have been following this, really appreciate all the eyes and feedback here. This PR fixes Mistral tool-call parsing. Since there are quite a few people depending on this, a code-owner review when you have a moment would mean a lot.  @chaunceyjiang @aarnphm 
Thanks again for all the help and for maintaining vLLM!

Blake-Martin-code commented:
> Hey folks. A bunch of us have been following this, really appreciate all the eyes and feedback here. This PR fixes Mistral tool-call parsing. Since there are quite a few people depending on this, a code-owner review when you have a moment would mean a lot. @chaunceyjiang @aarnphm Thanks again for all the help and for maintaining vLLM!

Any updates on this? Very excited for this to go through

avigny commented:
> Also question. Does this merge request handle the issue where you cannot use mistral if you try to override the chat template. I get an error when trying to use the parallel jinja chat template. Says the tokenizer does not let you override the default chat template.

No it does not. I believe this is a choice from Mistral and not a bug from vllm

LittleCoinCoin commented:
Hello there,
Just pinging to keep the thread alive: I would also welcome this merge for my use cases :)

avigny committed:
Merge remote-tracking branch 'upstream/main' into mistral-tool-parser-streaming-update

# Conflicts:
#	vllm/entrypoints/openai/tool_parsers/mistral_tool_parser.py

avigny committed:
Merge branch 'main' into mistral-tool-parser-streaming-update

avigny committed:
ruff repair

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

chaunceyjiang reviewed (approved):
I'm very sorry I missed this PR. Let's merge it quickly so we can promptly add support for Ministral-3.

gaby commented:
@chaunceyjiang Thank you!

chaunceyjiang commented on tests/tool_use/test_mistral_tool_parser.py:
```diff
@@ -0,0 +1,839 @@
+from mistral_common.protocol.instruct.request import InstructRequest
+from mistral_common.protocol.instruct.tool_calls import FunctionCall, ToolCall
+from partial_json_parser.core.options import Allow
+from vllm.transformers_utils.detokenizer_utils import detokenize_incrementally
```
```suggestion
from vllm.tokenizers.detokenizer_utils import detokenize_incrementally
```

chaunceyjiang committed:
Apply suggestion from @chaunceyjiang

Signed-off-by: Chauncey <chaunceyjiang@gmail.com>

chaunceyjiang commented on tests/tool_use/test_mistral_tool_parser.py:
```diff
@@ -0,0 +1,839 @@
+from mistral_common.protocol.instruct.request import InstructRequest
+from mistral_common.protocol.instruct.tool_calls import FunctionCall, ToolCall
+from partial_json_parser.core.options import Allow
+from vllm.tokenizers.detokenizer_utils import detokenize_incrementally
```
```suggestion

```

chaunceyjiang committed:
Fix format error in tool parser

Signed-off-by: chaunceyjiang <chaunceyjiang@gmail.com>

DarkLight1337 committed:
Merge branch 'main' into mistral-tool-parser-streaming-update

avigny committed:
Proper use of `mistral_tokenizer.version` for `stream_delta_message_generator`

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny committed:
[unit tests] Adding a space in the model output

Signed-off-by: avigny <47987522+avigny@users.noreply.github.com>

avigny commented:
> @avigny just tested the new mistralai/Ministral-3-14B-Instruct-2512 , tool calling with non streaming works. streaming tool calling still doesn't ... any chance this fix also solves that for the new model or is specific to Mistral3.2 Small?

The new `mistralai/Ministral-3-14B-Instruct-2512` for example is using a mistral tokenizer v13 (see the [tekken](https://huggingface.co/mistralai/Ministral-3-14B-Instruct-2512/blob/main/tekken.json) file) and this implies yet another format, different from the ones repaired by this PR.
This PR does not bring tool calls for the new mistral models.

alew3 commented:
> > @avigny just tested the new mistralai/Ministral-3-14B-Instruct-2512 , tool calling with non streaming works. streaming tool calling still doesn't ... any chance this fix also solves that for the new model or is specific to Mistral3.2 Small?
> 
> The new `mistralai/Ministral-3-14B-Instruct-2512` for example is using a mistral tokenizer v13 (see the [tekken](https://huggingface.co/mistralai/Ministral-3-14B-Instruct-2512/blob/main/tekken.json) file) and this implies yet another format (from what I recall), different from the ones repaired by this PR. Haven't tested these new models but I believe this PR does not bring tool calls for the new mistral models.



> > @avigny just tested the new mistralai/Ministral-3-14B-Instruct-2512 , tool calling with non streaming works. streaming tool calling still doesn't ... any chance this fix also solves that for the new model or is specific to Mistral3.2 Small?
> 
> The new `mistralai/Ministral-3-14B-Instruct-2512` for example is using a mistral tokenizer v13 (see the [tekken](https://huggingface.co/mistralai/Ministral-3-14B-Instruct-2512/blob/main/tekken.json) file) and this implies yet another format (from what I recall), different from the ones repaired by this PR. Haven't tested these new models but I believe this PR does not bring tool calls for the new mistral models.

That's a shame, let's hope it won't be as complicated as this fix. I wish Mistral would release the new models with tool calling working on day 0. Most use cases now a days are pretty much useless without it. Thanks for all your efforts and persistence!

avigny committed:
Merge branch 'main' into mistral-tool-parser-streaming-update

avigny committed:
Merge branch 'main' into mistral-tool-parser-streaming-update

avigny commented:
@Erwandsn this looks like a tool call generated by a v13 tokenizer, which is not handled by this PR.
The more recent models do have tool call format a bit different from the previous models.
This issue is tracked:
#23180
#29968

graelo commented:
Hello, I'll soon open a PR with a fix for this that relies on mistral-common (with MistralTokenizer). There will be 2 commits:
- the first supports both pre-v11 and v11+ tokenizers,
- the second commit removes support for pre-v11 models such as mistral-7b, because quite frankly the tool call support for these older models is quite limited. It simplifies the code. I'll let the conversation decide.

If you want to test it in the meantime (I'd be glad to know it also works for others), here's the updated code (from `main`, after this PR was merged yesterday). I've been using `mistralai/ministral-3-14b-instruct-2512` without an issue with this.

<details>
This is how I mount this for local tests, I use `docker.io/vllm/vllm-openai:latest` (so v0.12.0 from yesterday):

```bash
--volume=/path/to/local/mistral_tool_parser.py:/usr/local/lib/python3.12/dist-packages/vllm/entrypoints/openai/tool_parsers/mistral_tool_parser.py
```

Here is the updated code for my upcoming PR.

```python
# SPDX-License-Identifier: Apache-2.0
# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
"""
Mistral tool call parser for v11+ models.

This implementation uses token-based parsing for streaming, leveraging the
atomic nature of special token IDs ([TOOL_CALLS], [ARGS], [CALL_ID]) to
reliably detect tool call boundaries.

Supported models: Mistral-Small-3.1+, Ministral-3+, and other v11+ models.

Note: Pre-v11 models (Mistral-7B-Instruct-v0.1/v0.2/v0.3) are not supported.
These older models have limited tool calling capabilities and require complex
text-based parsing with partial JSON handling. Users should upgrade to v11+
models for reliable tool calling support.
"""

import json
from collections.abc import Sequence
from enum import Enum, auto
from random import choices
from string import ascii_letters, digits

import regex as re
from pydantic import Field

from vllm.entrypoints.openai.protocol import (
    ChatCompletionRequest,
    DeltaFunctionCall,
    DeltaMessage,
    DeltaToolCall,
    ExtractedToolCallInformation,
    FunctionCall,
    ToolCall,
)
from vllm.entrypoints.openai.tool_parsers.abstract_tool_parser import (
    ToolParser,
)
from vllm.logger import init_logger
from vllm.tokenizers import MistralTokenizer, TokenizerLike

logger = init_logger(__name__)

ALPHANUMERIC = ascii_letters + digits


class MistralToolCall(ToolCall):
    id: str = Field(default_factory=lambda: MistralToolCall.generate_random_id())

    @staticmethod
    def generate_random_id():
        # Mistral Tool Call Ids must be alphanumeric with a length of 9.
        # https://github.com/mistralai/mistral-common/blob/21ee9f6cee3441e9bb1e6ed2d10173f90bd9b94b/src/mistral_common/protocol/instruct/validator.py#L299
        return "".join(choices(ALPHANUMERIC, k=9))

    @staticmethod
    def is_valid_id(id: str) -> bool:
        return id.isalnum() and len(id) == 9


class StreamingState(Enum):
    """Streaming state for tool call parsing."""

    CONTENT = auto()  # Before any [TOOL_CALLS] token
    PARSING_TOOL_NAME = auto()  # After [TOOL_CALLS], parsing function name
    PARSING_TOOL_ARGS = auto()  # Parsing JSON arguments
    COMPLETE = auto()  # All tools parsed


class MistralToolParser(ToolParser):
    """
    Tool call parser for Mistral v11+ models.

    Supports the v11+ format: [TOOL_CALLS]name[ARGS]{...}
    Optionally with call ID: [TOOL_CALLS]name[CALL_ID]id[ARGS]{...}

    This parser requires MistralTokenizer (tokenizer_mode=mistral) and
    models using tokenizer version 11 or higher.
    """

    def __init__(self, tokenizer: TokenizerLike):
        super().__init__(tokenizer)

        if not isinstance(self.model_tokenizer, MistralTokenizer):
            raise RuntimeError(
                "MistralToolParser requires MistralTokenizer. "
                "Please use tokenizer_mode='mistral' in your vLLM configuration. "
                "Note: Only v11+ Mistral models are supported for tool calling."
            )

        self._mistral_base_tokenizer = self.model_tokenizer.tokenizer
        self._version = self.model_tokenizer.version

        if self._version < 11:
            raise RuntimeError(
                f"MistralToolParser requires tokenizer version 11 or higher, "
                f"but got version {self._version}. Pre-v11 models "
                "(Mistral-7B-Instruct-v0.1/v0.2/v0.3) are not supported for "
                "tool calling. Please use a v11+ model such as "
                "Mistral-Small-3.1 or Ministral-3."
            )

        # Get bot token info
        self.bot_token = "[TOOL_CALLS]"
        self.bot_token_id = self.vocab.get(self.bot_token)

        if self.bot_token_id is None:
            raise RuntimeError(
                "Mistral Tool Parser could not locate the [TOOL_CALLS] token "
                "in the tokenizer!"
            )

        # Get control tokens for v11+ format
        try:
            self._args_token_id = self._mistral_base_tokenizer.get_control_token(
                "[ARGS]"
            )
        except Exception:
            raise RuntimeError(
                "Mistral Tool Parser could not locate the [ARGS] token. "
                "This token is required for v11+ tool call parsing."
            )

        self._call_id_token_id: int | None = None
        try:
            self._call_id_token_id = self._mistral_base_tokenizer.get_control_token(
                "[CALL_ID]"
            )
        except Exception:
            # [CALL_ID] is optional - some models may not have it
            pass

        # Regex for non-streaming parsing: name{args}
        self.fn_name_regex = re.compile(
            r"([a-zA-Z0-9_-]+)(\{[\s\S]*?\}+)", re.DOTALL
        )

        # Streaming state
        self._streaming_state = StreamingState.CONTENT
        self._current_tool_index = -1
        self._current_tool_id: str | None = None
        self._current_tool_name: str = ""
        self._current_tool_args: str = ""
        self._brace_depth = 0

        # For compatibility with serving_chat.py's finish_reason detection
        self.prev_tool_call_arr: list[dict] = []

    def extract_tool_calls(
        self,
        model_output: str,
        request: ChatCompletionRequest,
    ) -> ExtractedToolCallInformation:
        """
        Extract tool calls from a complete model response.

        Parses the v11+ format: [TOOL_CALLS]name{args}[TOOL_CALLS]name{args}...
        """
        # Fast path: no tool call token present
        if self.bot_token not in model_output:
            return ExtractedToolCallInformation(
                tools_called=False, tool_calls=[], content=model_output
            )

        try:
            # Get content before tool calls
            content = model_output.split(self.bot_token)[0]
            content = content if content.strip() else None

            # Parse tool calls from each segment after [TOOL_CALLS]
            function_call_arr = []
            for segment in model_output.split(self.bot_token):
                if not segment.strip():
                    continue

                matches = self.fn_name_regex.findall(segment)
                for match in matches:
                    fn_name = match[0]
                    args = match[1]
                    function_call_arr.append(
                        {"name": fn_name, "arguments": json.loads(args)}
                    )

            # Convert to MistralToolCall objects
            tool_calls: list[MistralToolCall] = [
                MistralToolCall(
                    type="function",
                    function=FunctionCall(
                        name=raw_function_call["name"],
                        arguments=json.dumps(
                            raw_function_call["arguments"], ensure_ascii=False
                        ),
                    ),
                )
                for raw_function_call in function_call_arr
            ]

            return ExtractedToolCallInformation(
                tools_called=True,
                tool_calls=tool_calls,
                content=content,
            )

        except Exception:
            logger.exception("Error in extracting tool call from response.")
            return ExtractedToolCallInformation(
                tools_called=False,
                tool_calls=[],
                content=model_output.replace(self.bot_token, "").strip(),
            )

    def extract_tool_calls_streaming(
        self,
        previous_text: str,
        current_text: str,
        delta_text: str,
        previous_token_ids: Sequence[int],
        current_token_ids: Sequence[int],
        delta_token_ids: Sequence[int],
        request: ChatCompletionRequest,
    ) -> DeltaMessage | None:
        """
        Extract tool calls from streaming output using token-based parsing.

        Token IDs are atomic - they cannot be split across chunks - which
        eliminates a whole class of parsing bugs that affect text-based parsing.
        """
        # If no tool call token seen yet, emit as content
        if self.bot_token_id not in current_token_ids:
            return DeltaMessage(content=delta_text)

        return self._stream_tool_calls(delta_token_ids)

    def _stream_tool_calls(
        self, delta_token_ids: Sequence[int]
    ) -> DeltaMessage | None:
        """
        Stream tool calls using token-based parsing.

        Detects [TOOL_CALLS] and [ARGS] tokens to identify tool call boundaries,
        then streams function names and arguments as they arrive.
        """
        from mistral_common.tokens.tokenizers.base import SpecialTokenPolicy

        delta_tool_calls: list[DeltaToolCall] = []

        for token_id in delta_token_ids:
            if token_id == self.bot_token_id:
                # Starting a new tool call
                self._current_tool_index += 1
                self._current_tool_id = MistralToolCall.generate_random_id()
                self._current_tool_name = ""
                self._current_tool_args = ""
                self._brace_depth = 0
                self._streaming_state = StreamingState.PARSING_TOOL_NAME

                # Set flag for finish_reason detection
                if not self.prev_tool_call_arr:
                    self.prev_tool_call_arr = [{"arguments": {}}]

                # Initialize streamed_args_for_tool for this tool index
                while len(self.streamed_args_for_tool) <= self._current_tool_index:
                    self.streamed_args_for_tool.append("")

            elif token_id == self._args_token_id:
                # Transition from name to arguments
                if self._streaming_state == StreamingState.PARSING_TOOL_NAME:
                    # Emit the complete function name
                    delta_tool_calls.append(
                        DeltaToolCall(
                            index=self._current_tool_index,
                            type="function",
                            id=self._current_tool_id,
                            function=DeltaFunctionCall(
                                name=self._current_tool_name.strip()
                            ).model_dump(exclude_none=True),
                        )
                    )
                    self._streaming_state = StreamingState.PARSING_TOOL_ARGS

            elif token_id == self._call_id_token_id:
                # Skip call ID tokens (they come between name and [ARGS])
                # We generate our own IDs
                pass

            elif self._streaming_state == StreamingState.CONTENT:
                # Before any tool call - shouldn't happen if bot_token_id
                # is in current_token_ids, but handle gracefully
                pass

            elif self._streaming_state == StreamingState.PARSING_TOOL_NAME:
                # Accumulate name tokens
                token_str = self._mistral_base_tokenizer.decode(
                    [token_id], special_token_policy=SpecialTokenPolicy.IGNORE
                )
                self._current_tool_name += token_str

            elif self._streaming_state == StreamingState.PARSING_TOOL_ARGS:
                # Stream argument tokens
                token_str = self._mistral_base_tokenizer.decode(
                    [token_id], special_token_policy=SpecialTokenPolicy.IGNORE
                )

                # Track brace depth for nested JSON
                for char in token_str:
                    if char == "{":
                        self._brace_depth += 1
                    elif char == "}":
                        self._brace_depth -= 1

                self._current_tool_args += token_str

                # Update streamed_args_for_tool for vLLM's finish handling
                if self._current_tool_index < len(self.streamed_args_for_tool):
                    self.streamed_args_for_tool[self._current_tool_index] = (
                        self._current_tool_args
                    )

                # Emit arguments delta
                delta_tool_calls.append(
                    DeltaToolCall(
                        index=self._current_tool_index,
                        function=DeltaFunctionCall(
                            arguments=token_str
                        ).model_dump(exclude_none=True),
                    )
                )

        # Build response
        if delta_tool_calls:
            return DeltaMessage(tool_calls=delta_tool_calls)

        return None
```
</details>

PS: I had already followed this approach in a PR on mlx-lm this summer. The `mistral-common` library is underrated.

Erwandsn commented:
@graelo I just tested with the latest code, and stil the same issue but 3 lines before:

<details>

<summary>Error output</summary>

`(APIServer pid=1) ERROR 12-06 20:28:15 [mistral_tool_parser.py:202]   File "/app/.venv/lib/python3.12/site-packages/vllm/entrypoints/openai/tool_parsers/mistral_tool_parser.py", line 178, in extract_tool_calls
(APIServer pid=1) ERROR 12-06 20:28:15 [mistral_tool_parser.py:202]     {"name": fn_name, "arguments": json.loads(args)}
(APIServer pid=1) ERROR 12-06 20:28:15 [mistral_tool_parser.py:202]                                    ^^^^^^^^^^^^^^^^
(APIServer pid=1) ERROR 12-06 20:28:15 [mistral_tool_parser.py:202]   File "/usr/lib/python3.12/json/__init__.py", line 346, in loads
(APIServer pid=1) ERROR 12-06 20:28:15 [mistral_tool_parser.py:202]     return _default_decoder.decode(s)
(APIServer pid=1) ERROR 12-06 20:28:15 [mistral_tool_parser.py:202]            ^^^^^^^^^^^^^^^^^^^^^^^^^^
(APIServer pid=1) ERROR 12-06 20:28:15 [mistral_tool_parser.py:202]   File "/usr/lib/python3.12/json/decoder.py", line 337, in decode
(APIServer pid=1) ERROR 12-06 20:28:15 [mistral_tool_parser.py:202]     obj, end = self.raw_decode(s, idx=_w(s, 0).end())
(APIServer pid=1) ERROR 12-06 20:28:15 [mistral_tool_parser.py:202]                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(APIServer pid=1) ERROR 12-06 20:28:15 [mistral_tool_parser.py:202]   File "/usr/lib/python3.12/json/decoder.py", line 353, in raw_decode
(APIServer pid=1) ERROR 12-06 20:28:15 [mistral_tool_parser.py:202]     obj, end = self.scan_once(s, idx)
(APIServer pid=1) ERROR 12-06 20:28:15 [mistral_tool_parser.py:202]                ^^^^^^^^^^^^^^^^^^^^^^
(APIServer pid=1) ERROR 12-06 20:28:15 [mistral_tool_parser.py:202] json.decoder.JSONDecodeError: Unterminated string starting at: line 1 column 44 (char 43)`

</details>

Erwandsn commented:
@avigny No problem, this is really a weird issue. I also think this is related with the complexity of the json (wich doesn't help for debug). 

But i had the case with smaller json, i think the complexity of variables escape when HTML is included in the JSON make the model mess up, (I'm on Ministral-3-8b-instruct). 

And my way to pass files content with JSON really suck and is a part of the problem. So i think this is not an emergency issue. 

Thanks for your investigations !
